Add a command line tool for load testing the Provisioner

gigawhitlocks opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
This PR introduces a new tool in the `cmd/tools` directory which can be installed with `go install ./cmd/tools/cloudburst` from the project base directory.

`cloudburst` is a command line tool for load testing the Provisioner. `cloudburst` allows you to very quickly spin up a huge number of Installations and collects timing data as it runs.

```
Run a load test against a provisioning server.

Usage:
  cloudburst [flags]

Flags:
      --batch int       Specify the number of Installations in each batch. Installations in a batch are install serially and batches are installed in parallel. (default 5)
  -h, --help            help for cloudburst
      --runs int        Number of times to repeat the test (default 1)
      --server string   Location of the Provisioning Server to load test (default "http://localhost:8075")
      --total int       Number of Installations to provision (default 20)
```

The nature of this PR is such that this code won't get "stale" the way a more tightly integrated PR might, so feel free to make larger requests in this PR than normal, as my primary goal here is to provide a tool that has an immediate as well as long term value payoff.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
No ticket; addresses "explosive scaling testing" OKR goal.

#### Release Note
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

-->

```release-note
Added cloudburst, a tool for explosive scaling / load testing.
```

gigawhitlocks commented on cmd/tools/cloudburst/cloudburst.go:
```diff
@@ -0,0 +1,409 @@
+		blaster.waitForGroup()
+
+		allReports := []Report{}
+		for i := 0; i < runs; i++ {
```
Forgot to refactor this bit before opening the PR. Will do that.

gabrieljackson commented on cmd/tools/cloudburst/cloudburst.go:
```diff
@@ -0,0 +1,414 @@
+		return
+	}
+}
+func init() {
```
nit: `init()` near the top of the file would be nice.

gabrieljackson commented on cmd/tools/cloudburst/cloudburst.go:
```diff
@@ -0,0 +1,414 @@
+	logger = log.New()
+	blastCommand.PersistentFlags().String("server", "http://localhost:8075", "Location of the Provisioning Server to load test")
+	blastCommand.PersistentFlags().Int("runs", 1, "Number of times to repeat the test")
+	blastCommand.PersistentFlags().Int("batch", 5, "Specify the number of Installations in each batch. Installations in a batch are install serially and batches are installed in parallel.")
```
```suggestion
	blastCommand.PersistentFlags().Int("batch", 5, "Specify the number of Installations in each batch. Installations in a batch are installed serially and batches are installed in parallel.")
```

gabrieljackson commented on cmd/tools/cloudburst/cloudburst.go:
```diff
@@ -0,0 +1,414 @@
+// returns an error if the API does. It does not block or retry.
+func (b *Blaster) createInstallation() (*cloud.Installation, error) {
+	installationDTO, err := b.client.CreateInstallation(
+		&cloud.CreateInstallationRequest{
```
Being able to specify `version`, `filestore`, and `database` with flags in the future would be nice.

gigawhitlocks replied:
https://mattermost.atlassian.net/browse/MM-30529 Created this story to track this work in the future

gabrieljackson reviewed (approved):
Very cool. This will definitely be helpful to have around.

Szymongib commented on cmd/tools/cloudburst/cloudburst.go:
```diff
@@ -0,0 +1,414 @@
+		}
+
+		logger.Infof("Waiting for Group to be created..")
+		blaster.waitForGroup()
```
Isn't the group creation synchronous operation? I am not sure if we need to wait for a group to be created.

gigawhitlocks replied:
It is not synchronous however the aggressive retrial strategy in the routines that send `CreateInstallationRequest`s does probably make this redundant. I will double-check and remove this if I find that it is not necessary.

gigawhitlocks commented:
Sorry about the wrecked history on this PR. I moved `init()` to satisfy one of Gabe's comments and rebased against `master` instead of merging, without thinking. Bleh.
